### PR TITLE
Bug 2086728: Move node draining to controller. Remove MCD privs

### PR DESCRIPTION
**Pull request by yuqi-zhang**
This is PoC/testing stage currently. This will work (mostly as it will today)

1. Add a new drain controller that has 5 workers processing work queue separately in 90s drain intervals for requesting nodes, to create some simple parallelism
  a) This drain controller preserves some state in terms of when the drain started, and fails after 1 hour. This state (and any ongoing drains) will be lost when the controller restarts, but the daemon will still preserve the failure timing and write errors as we used to. The eviction api will still be processing pod terminations in the meantime (I think)
  b) If you have > 5 nodes draining in parallel, other nodes will start processing after the 90 second mark, so technically the parallelism extends (but at a slower pace)
  c) The controller pod itself no longer gets drained (since it will need to drain its own node) when that node updates, and instead just gets sigterm/sigkill'ed when the node is rebooting. This will cause drain delay for other nodes as the controller gets churned around (re-starting, leader election, etc.) compared to individual node pods draining themselves
  d) This new controller coordinates with daemons via two new annotations: desiredDrain and lastAppliedDrain. This isn't exactly bulletproof
  e) Error modes have changed from the daemon to the controller (daemon still have errors as we do today, but the details are in the controller now)

2. The daemon today just swaps in any drain operation with a request-via-annotation and then a blocking wait. This logic needs to be hardened for state-machine based logic in case of daemon pod restarts (but it mostly matches the behaviour today).
In terms of privs, we can (most likely) use kubelet kubeconfig, but we also would need to read/write MCs, which are a cluster scoped CR today. Short of reworking that (into a mountable configmap, for example), our options are:
  a) remove some perms, and keep node perms + MCs (although this does mean that we can write control plane nodes with our RBAC, which probably isn't ideal)
  b) have two sets of clients, one for reading/writing MCs, one for everything else (using the kubelet kubeconfig)

3. As a longer term consideration, we should probably be moving drain/reboot coordination outside of the MCO. We should have a disruption-coordination-operator of sorts, which takes drain and reboot requests via CRDs. For the MCO and other operators (SRIOV, intel, NVIDIA, etc.) this can act as a server side drain operation as well as reboot coordinator to ensure we are not overwriting each other, and the individual operators would no longer need to handle their own drains/reboots.

For other considerations and Hypershift, see the Hypershift google doc.

### CODE REVIEW

File: pkg/controller/drain/drain_controller.go [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+	}()
+
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	// First check if a drain request is needed for this node
+	// TODO (jerzhang): consider not using the cache for this,
+	// since it has been shown to be problematic in the past
```
Comment (cgwalters) [inline pkg/controller/drain/drain_controller.go:219]: Curious about that

File: pkg/controller/drain/drain_controller.go [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+	desiredState := node.Annotations[daemonconsts.DesiredDrainerAnnotationKey]
+	if desiredState == node.Annotations[daemonconsts.LastAppliedDrainerAnnotationKey] {
+		glog.V(4).Infof("Node %v has the correct drain", key)
+		return nil
+	}
+
+	podName, ok := os.LookupEnv("POD_NAME")
+	if !ok || podName == "" {
+		glog.Fatalf("pod name is required for controller drainer")
+	}
```
Comment (cgwalters) [inline pkg/controller/drain/drain_controller.go:238]: I think we could do this once at startup and cache it.

File: pkg/controller/drain/drain_controller.go [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+	desiredVerb := strings.Split(desiredState, "-")[0]
+	if desiredVerb == daemonconsts.DrainerStateUncordon {
+		glog.Infof("node %v to be uncordoned", key)
+		// perform uncordon
+		if err := ctrl.cordonOrUncordonNode(false, node, drainer); err != nil {
+			return fmt.Errorf("failed to uncordon node %v: %w", node.Name, err)
+		}
+	}
+
+	if desiredVerb == daemonconsts.DrainerStateDrain {
```
Comment (cgwalters) [inline pkg/controller/drain/drain_controller.go:276]: This could be an `else if` or a `switch`.

File: pkg/controller/drain/drain_controller.go [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+		// First check if we have an ongoing drain
+		// This is currently stored in the object itself as a map but,
+		// Practically during upgrades the control plane node this controller
+		// pod is running on will also be terminated (the drainer will skip it).
+		// This is a bit problematic in practice since we don't really have a previous state.
+		// TODO (jerzhang) consider using a new CRD for coordination
```
Comment (cgwalters) [inline pkg/controller/drain/drain_controller.go:282]: Hmm.  What if the Node object had an ObjectReference to the thing doing the drain?  The benefit of that is it'd become a clear central point of knowledge/coordination.

Short term maybe we could shoehorn such a reference into an annotation.

For example, suppose that this drain pod is killed unexpectedly (control plane node went down, it hit OOM or whatever).

The pod restarts on another host (or same host in OOM scenario after OOM clears out), sees the annotation, and checks whether the pod that it references is still active or not.

If not, it knows to reset the annotation.  Otherwise, we should skip trying to drain and assume the other pod has it?

File: pkg/controller/drain/drain_controller.go [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+
+		glog.Infof("node %v to be drained", key)
+		ongoingDrain := false
+		for k, v := range ctrl.ongoingDrains {
+			if k == node.Name {
+				glog.Infof("TEST: previous node drain found. Node %v time %v", k, v)
+				ongoingDrain = true
+				duration := time.Now().Sub(v).Hours()
+				glog.Infof("Drain has been going on for %v hours", duration)
+				if duration > 1 {
```
Comment (cgwalters) [inline pkg/controller/drain/drain_controller.go]: I think it's worth hoisting this up as a clear `const` at the top.

File: pkg/controller/drain/drain_controller.go [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+					return fmt.Errorf("Drain has exceeded the 1 hour error mark")
```
Comment (cgwalters) [inline pkg/controller/drain/drain_controller.go]: Then avoid repeating it with e.g.:
`return fmt.Errorf("Drain exceeded timeout: %v", drainMaxDuration)`
or so

File: pkg/controller/drain/drain_controller.go [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+			if err := ctrl.cordonOrUncordonNode(true, node, drainer); err != nil {
+				return fmt.Errorf("failed to uncordon node %v: %w", node.Name, err)
+			}
+			ctrl.ongoingDrains[node.Name] = time.Now()
+		}
+
+		// Attempt drain
+		if err := drain.RunNodeDrain(drainer, node.Name); err != nil {
+			glog.Infof("Drain failed, but overall timeout has not been reached. Waiting 1 minute then retrying. Error message from drain: %v", err)
+			ctrl.enqueueAfter(node, 1*time.Minute)
```
Comment (cgwalters) [inline pkg/controller/drain/drain_controller.go:312]: More arbitrary times I think would be good as `const`.

**Review by cgwalters — COMMENTED** [review]
Awesome, thanks for starting this WIP!

One thing I think would be useful to think about is - can we make this somehow an opt-in switch?  That would make it much easier to experiment with and avoid a "flag day".

(The downside of such a thing is we'd need to re-roll the MCD permissions when it changes, etc.)

File: pkg/controller/drain/drain_controller.go (hunk 1 above)
Comment (yuqi-zhang) [reply pkg/controller/drain/drain_controller.go:219]: Our current uncordon logic doesn't actually work very well for 2 reasons:
1. minor bug in our logic that we remove the file before success, so if we timeout the first sync for some reason, we never uncordon
2. the uncordon checks we have don't work since the cache isnt updated fast enough (via nodelister) so we are always reading a stale status (on whether the node is successfully uncordoned or not)

File: pkg/controller/drain/drain_controller.go (hunk 2 above)
Comment (yuqi-zhang) [reply pkg/controller/drain/drain_controller.go:238]: Sure, I can cache it into the controller object itself

File: pkg/controller/drain/drain_controller.go (hunk 3 above)
Comment (yuqi-zhang) [reply pkg/controller/drain/drain_controller.go:276]: I'll switch to a switch :)

File: pkg/controller/drain/drain_controller.go (hunk 4 above)
Comment (yuqi-zhang) [reply pkg/controller/drain/drain_controller.go:282]: As in, so long as we started the drain and got to the eviction API call, it's technically always ongoing.

**Comment by yuqi-zhang** [comment]
I can't reproduce the failure locally, even running the e2e test directly. Flake or not?

/test e2e-gcp-op

File: pkg/controller/drain/drain_controller.go (hunk 4 above)
Comment (cgwalters) [reply pkg/controller/drain/drain_controller.go:282]: Or to rephrase, while in Go we're making *one* API call, it internally is implemented with many requests to the API server (see `evictPods()`) and hence is not transactional, so we need to handle being interrupted in the middle and restarting.

**Comment by cgwalters** [comment]
A few discrete tasks:

- One option is to make this code dynamically switchable.  Jerry and I were chatting about this, I think it's basically we have a global config option (a configmap?  annotation on something?; maybe strawman `oc create -n openshift-machine-config-operator configmap centralized-drain`) that enables this (or disables it) - this reduces the risk of landing.  To implement this the operator `sync.go` code updates the RBAC and also the daemonset to pass e.g. `--centralized-drain` in between reading manifest and applying.
- Actually switch the daemon to use kubelet credentials
- Drop write on machineconfig
- Split out smaller prep patches like https://github.com/openshift/machine-config-operator/pull/3135#discussion_r865912834
- Fix config drift monitor race start time racing with subsequent updates; this seems to relate to Degraded state potentially?  [zack is taking this]

Followups:

- Pick up fix for leader election

**Comment by cgwalters** [comment]
Ah yes, I forgot to start the informer; classic.  Will look at fixing that up.

**Comment by cgwalters** [comment]
OK I have https://github.com/openshift/machine-config-operator/pull/3141 in a rough (I think) working shape.

File: pkg/daemon/daemon.go [hunk 8]
```diff
@@ -1716,8 +1694,27 @@ func (dn *Daemon) prepUpdateFromCluster() (*mcfgv1.MachineConfig, *mcfgv1.Machin
 // "transient state" file, which signifies that all of those prior steps have
 // been completed.
 func (dn *Daemon) completeUpdate(desiredConfigName string) error {
-	if err := dn.cordonOrUncordonNode(false); err != nil {
-		return err
+	if err := dn.nodeWriter.SetDesiredDrainer(dn.kubeClient.CoreV1().Nodes(), dn.nodeLister, dn.name, fmt.Sprintf("%s-%s", "uncordon", desiredConfigName)); err != nil {
+		return fmt.Errorf("Could not set drain annotation: %w", err)
+	}
+
```
Comment (cheesesashimi) [inline pkg/daemon/daemon.go:1700]: Lets add a call to `dn.startConfigDriftMonitor()` here. This will ensure that the Config Drift Monitor is active before we uncordon the node.

**Comment by yuqi-zhang** [comment]
Rebased onto https://github.com/openshift/machine-config-operator/pull/3141 for testing purposes

**Comment by yuqi-zhang** [comment]
The final set of perms for the MCD would then break down as:

```
- apiGroups: [""]
  resources: ["nodes"]
  verbs: ["get", "list", "watch"]
```
To react to node object changes
```
- apiGroups: ["machineconfiguration.openshift.io"]
  resources: ["machineconfigs"]
  verbs: verbs: ["get", "list", "watch"]
```
To fetch the necessary MCs
```
- apiGroups: ["security.openshift.io"]
  resourceNames: ["privileged"]
  resources: ["securitycontextconstraints"]
  verbs: ["use"]
```
For host networking
```
- apiGroups:
  - authentication.k8s.io
  resources:
  - tokenreviews
  - subjectaccessreviews
  verbs:
  - create
- apiGroups:
  - authorization.k8s.io
  resources:
  - subjectaccessreviews
  verbs:
  - create
```
For prom metrics

**Comment by yuqi-zhang** [comment]
The gcp-op failures will be separately addressed via https://github.com/openshift/machine-config-operator/pull/3146

/test all

**Comment by yuqi-zhang** [comment]
Rebased and addressed comments.

gcp-op is blocked on https://github.com/openshift/installer/pull/5898 but tested to work fine locally still. Ready for another round

File: pkg/controller/drain/drain_controller.go [hunk 9]
```diff
@@ -0,0 +1,425 @@
...
+	if apierrors.IsNotFound(err) {
+		glog.V(2).Infof("node %v has been deleted", key)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	desiredState := node.Annotations[daemonconsts.DesiredDrainerAnnotationKey]
+	if desiredState == node.Annotations[daemonconsts.LastAppliedDrainerAnnotationKey] {
```
Comment (cgwalters) [inline pkg/controller/drain/drain_controller.go:241]: One thing I found confusing here is the case of "newly joined node" (initial cluster install or scaled up).  What I expected to see is code around here that says "no drain requested".

But it looks like we end dropping down into...the "uncordon" flow?

Comment (yuqi-zhang) [reply pkg/controller/drain/drain_controller.go:241]: Ah yeah there should be a nil check here, which hypershift has. I will add that on the next iteration

**Comment by cheesesashimi** [comment]
Just to clarify: https://github.com/openshift/machine-config-operator/pull/3146 will fix the Config Drift Monitor e2e tests. What was happening is the e2e test was naively looking for the presence of the Config Drift Monitor startup message in the MCD pod logs as a signal that it was running and ready to move onto the next test. Normally, the Config Drift Monitor always started back up rather quickly and this was never an issue. However, since this PR makes the process take a bit longer, we need to more definitively determine the state of the Config Drift Monitor. Additionally, the Config Drift Monitor e2e tests should determine if the node reboots during the process as we have documentation written around when the node reboots vs. doesn't. If that behavior were to change (intentionally or otherwise), the e2e test acts as a warning.

**Comment by sinnykumari** [comment]
Since, gcp-op test is green now, retrying
/test e2e-gcp-op

**Comment by cgwalters** [comment]
>  --- FAIL: TestRunShared (611.52s)
>    --- FAIL: TestRunShared/MCD_Config_Drift (611.52s) 

Presumably this needs https://github.com/openshift/machine-config-operator/pull/3146

**Comment by cgwalters** [comment]
OK I've rebased :surfer: on master and cherry picked in the commits from https://github.com/openshift/machine-config-operator/pull/3146/commits/

**Comment by cgwalters** [comment]
Does this patch look OK to push on top of this branch?

```
From 4b43c7a1ff7bdfd65e11a79ac16e782c40e00cc5 Mon Sep 17 00:00:00 2001
From: Colin Walters <walters@verbum.org>
Date: Mon, 16 May 2022 17:46:09 -0400
Subject: [PATCH] controller/drain: Clean up and make consistent log/error
 messages

I noticed when debugging a CI failure that some of the log messages
aren't prefixed with the affected node.  That's going to be crucial
information.

Add a `logNode` helper that is much like the `node_controller.go`'s
version in `logPoolNode` for similar reasons that consistently
emits `node <name>: ` as a prefix.

Change the errors we return to do the same although without a helper
function for now.
---
 pkg/controller/drain/drain_controller.go | 36 ++++++++++++++----------
 1 file changed, 21 insertions(+), 15 deletions(-)

diff --git a/pkg/controller/drain/drain_controller.go b/pkg/controller/drain/drain_controller.go
index 9e21daebe..191c4a9b0 100644
--- a/pkg/controller/drain/drain_controller.go
+++ b/pkg/controller/drain/drain_controller.go
@@ -135,6 +135,12 @@ func (ctrl *Controller) Run(workers int, stopCh <-chan struct{}) {
 	<-stopCh
 }
 
+// logNode emits a log message at informational level, prefixed with the node name in a consistent fashion.
+func (ctrl *Controller) logNode(node *corev1.Node, format string, args ...interface{}) {
+	msg := fmt.Sprintf(format, args...)
+	glog.Infof("node %s: %s", node.Name, msg)
+}
+
 func (ctrl *Controller) handleNodeEvent(node interface{}) {
 	n := node.(*corev1.Node)
 	glog.V(4).Infof("Updating Node %s", n.Name)
@@ -255,7 +261,7 @@ func (ctrl *Controller) syncNode(key string) error {
 			if usingEviction {
 				verbStr = "Evicted"
 			}
-			glog.Infof("Node: %s, %s pod %s/%s", node.Name, verbStr, pod.Namespace, pod.Name)
+			ctrl.logNode(node, "%s pod %s/%s", verbStr, pod.Namespace, pod.Name)
 		},
 		AdditionalFilters: []drain.PodFilter{
 			func(pod corev1.Pod) drain.PodDeleteStatus {
@@ -273,7 +279,7 @@ func (ctrl *Controller) syncNode(key string) error {
 	desiredVerb := strings.Split(desiredState, "-")[0]
 	switch desiredVerb {
 	case daemonconsts.DrainerStateUncordon:
-		glog.Infof("node %v to be uncordoned", key)
+		ctrl.logNode(node, "uncordoning")
 		// perform uncordon
 		if err := ctrl.cordonOrUncordonNode(false, node, drainer); err != nil {
 			return fmt.Errorf("failed to uncordon node %v: %w", node.Name, err)
@@ -286,7 +292,7 @@ func (ctrl *Controller) syncNode(key string) error {
 		// This is a bit problematic in practice since we don't really have a previous state.
 		// TODO (jerzhang) consider using a new CRD for coordination
 
-		glog.Infof("node %v to be drained", key)
+		ctrl.logNode(node, "initiating drain")
 		ongoingDrain := false
 		for k, v := range ctrl.ongoingDrains {
 			if k != node.Name {
@@ -297,13 +303,13 @@ func (ctrl *Controller) syncNode(key string) error {
 			glog.Infof("Previous node drain found. Drain has been going on for %v hours", duration.Hours())
 			if duration > drainTimeoutDuration {
 				// TODO (jerzhang): handle error better
-				return fmt.Errorf("Drain exceeded timeout: %v", drainTimeoutDuration)
+				return fmt.Errorf("node %s: drain exceeded timeout: %v", node.Name, drainTimeoutDuration)
 			}
 			break
 		}
 		if !ongoingDrain {
 			// TODO probably have this detect the updated status instead
-			glog.Infof("node %v to be cordoned", key)
+			ctrl.logNode(node, "cordoning")
 			// perform uncordon
 			if err := ctrl.cordonOrUncordonNode(true, node, drainer); err != nil {
 				return fmt.Errorf("failed to uncordon node %v: %w", node.Name, err)
@@ -313,7 +319,7 @@ func (ctrl *Controller) syncNode(key string) error {
 
 		// Attempt drain
 		if err := drain.RunNodeDrain(drainer, node.Name); err != nil {
-			glog.Infof("Drain failed, but overall timeout has not been reached. Waiting 1 minute then retrying. Error message from drain: %v", err)
+			ctrl.logNode(node, "Drain failed, but overall timeout has not been reached. Waiting 1 minute then retrying. Error message from drain: %v", err)
 			ctrl.enqueueAfter(node, drainRequeueDelay)
 			return nil
 		}
@@ -321,7 +327,7 @@ func (ctrl *Controller) syncNode(key string) error {
 		// Drain was successful. Delete the ongoing drain, then set the annotation
 		delete(ctrl.ongoingDrains, node.Name)
 	default:
-		return fmt.Errorf("non-recognized drain verb detected %s", desiredVerb)
+		return fmt.Errorf("node %s: non-recognized drain verb detected %s", node.Name, desiredVerb)
 	}
 
 	glog.Infof("Operation successful! Writing annotation then finishing.")
@@ -330,7 +336,7 @@ func (ctrl *Controller) syncNode(key string) error {
 		daemonconsts.LastAppliedDrainerAnnotationKey: desiredState,
 	}
 	if err := ctrl.setNodeAnnotations(node.Name, annotations); err != nil {
-		return fmt.Errorf("failed to set node uncordoned annotation for node %v: %w", node.Name, err)
+		return fmt.Errorf("node %s: failed to set node uncordoned annotation: %w", node.Name, err)
 	}
 
 	return nil
@@ -360,14 +366,14 @@ func (ctrl *Controller) setNodeAnnotations(nodeName string, annotations map[stri
 
 		patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldNode, newNode, corev1.Node{})
 		if err != nil {
-			return fmt.Errorf("failed to create patch for node %q: %w", nodeName, err)
+			return fmt.Errorf("node %s: failed to create patch for: %w", nodeName, err)
 		}
 
 		_, err = ctrl.kubeClient.CoreV1().Nodes().Patch(context.TODO(), nodeName, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
 		return err
 	}); err != nil {
 		// may be conflict if max retries were hit
-		return fmt.Errorf("unable to update node %s: %w", nodeName, err)
+		return fmt.Errorf("node %s: unable to update: %w", nodeName, err)
 	}
 	return nil
 }
@@ -389,7 +395,7 @@ func (ctrl *Controller) cordonOrUncordonNode(desired bool, node *corev1.Node, dr
 	if err := wait.ExponentialBackoff(backoff, func() (bool, error) {
 		// Log has been added to ensure that MCO is correctly performing cordon/uncordon.
 		// This should help us with debugging bugs like https://bugzilla.redhat.com/show_bug.cgi?id=2022387
-		glog.Infof("Initiating %s on node (currently schedulable: %t)", verb, !node.Spec.Unschedulable)
+		ctrl.logNode(node, "initiating %s (currently schedulable: %t)", verb, !node.Spec.Unschedulable)
 		err := drain.RunCordonOrUncordon(drainer, node, desired)
 		if err != nil {
 			lastErr = err
@@ -407,18 +413,18 @@ func (ctrl *Controller) cordonOrUncordonNode(desired bool, node *corev1.Node, dr
 
 		if node.Spec.Unschedulable != desired {
 			// See https://bugzilla.redhat.com/show_bug.cgi?id=2022387
-			glog.Infof("RunCordonOrUncordon() succeeded but node is still not in %s state, retrying", verb)
+			ctrl.logNode(node, "RunCordonOrUncordon() succeeded but node is still not in %s state, retrying", verb)
 			return false, nil
 		}
 
-		glog.Infof("%s succeeded on node (currently schedulable: %t)", verb, !updatedNode.Spec.Unschedulable)
+		ctrl.logNode(node, "%s succeeded (currently schedulable: %t)", verb, !updatedNode.Spec.Unschedulable)
 		return true, nil
 	}); err != nil {
 		if err == wait.ErrWaitTimeout {
 			errs := kubeErrs.NewAggregate([]error{err, lastErr})
-			return fmt.Errorf("failed to %s node (%d tries): %w", verb, backoff.Steps, errs)
+			return fmt.Errorf("node %s: failed to %s (%d tries): %w", node.Name, verb, backoff.Steps, errs)
 		}
-		return fmt.Errorf("failed to %s node: %w", verb, err)
+		return fmt.Errorf("node %s: failed to %s: %w", node.Name, verb, err)
 	}
 
 	return nil
-- 
2.35.1

```

**Comment by kikisdeliveryservice** [comment]
re: colin's suggestion above, I'm neutral - it seems like there are only a couple logs that are missing node name, so might be easier to just fix those if you want to keep this PR and commits on the smaller end.

**Comment by cgwalters** [comment]
Here's a smaller patch

```
diff --git a/pkg/controller/drain/drain_controller.go b/pkg/controller/drain/drain_controller.go
index 9e21daebe..6fe51377b 100644
--- a/pkg/controller/drain/drain_controller.go
+++ b/pkg/controller/drain/drain_controller.go
@@ -313,7 +313,7 @@ func (ctrl *Controller) syncNode(key string) error {
 
 		// Attempt drain
 		if err := drain.RunNodeDrain(drainer, node.Name); err != nil {
-			glog.Infof("Drain failed, but overall timeout has not been reached. Waiting 1 minute then retrying. Error message from drain: %v", err)
+			glog.Infof("node %s: Drain failed, but overall timeout has not been reached. Waiting 1 minute then retrying. Error message from drain: %v", node.Name, err)
 			ctrl.enqueueAfter(node, drainRequeueDelay)
 			return nil
 		}
@@ -389,7 +389,7 @@ func (ctrl *Controller) cordonOrUncordonNode(desired bool, node *corev1.Node, dr
 	if err := wait.ExponentialBackoff(backoff, func() (bool, error) {
 		// Log has been added to ensure that MCO is correctly performing cordon/uncordon.
 		// This should help us with debugging bugs like https://bugzilla.redhat.com/show_bug.cgi?id=2022387
-		glog.Infof("Initiating %s on node (currently schedulable: %t)", verb, !node.Spec.Unschedulable)
+		glog.Infof("Initiating %s on node %s (currently schedulable: %t)", node.Name, verb, !node.Spec.Unschedulable)
 		err := drain.RunCordonOrUncordon(drainer, node, desired)
 		if err != nil {
 			lastErr = err
@@ -407,7 +407,7 @@ func (ctrl *Controller) cordonOrUncordonNode(desired bool, node *corev1.Node, dr
 
 		if node.Spec.Unschedulable != desired {
 			// See https://bugzilla.redhat.com/show_bug.cgi?id=2022387
-			glog.Infof("RunCordonOrUncordon() succeeded but node is still not in %s state, retrying", verb)
+			glog.Infof("node %s: RunCordonOrUncordon() succeeded but node is still not in %s state, retrying", node.Name, verb)
 			return false, nil
 		}
 
@@ -416,9 +416,9 @@ func (ctrl *Controller) cordonOrUncordonNode(desired bool, node *corev1.Node, dr
 	}); err != nil {
 		if err == wait.ErrWaitTimeout {
 			errs := kubeErrs.NewAggregate([]error{err, lastErr})
-			return fmt.Errorf("failed to %s node (%d tries): %w", verb, backoff.Steps, errs)
+			return fmt.Errorf("failed to %s node %s (%d tries): %w", verb, node.Name, backoff.Steps, errs)
 		}
-		return fmt.Errorf("failed to %s node: %w", verb, err)
+		return fmt.Errorf("failed to %s node %s: %w", verb, node.Name, err)
 	}
 
 	return nil
```

So I can push that for now, but looking at the two I think I prefer the bigger fix to make the log messages all consistently formatted.  We're going to be debugging this and reading the log messages a lot, and being able to scan for `drain_controller: node %s` instead of having the node name sometimes at the start, sometimes in the middle, sometimes near the end is IMO a win worth a larger patch.

OTOH, a larger problem here is we're not really getting these error messages into the operator status.

**Comment by kikisdeliveryservice** [comment]
second patch WFM, might be worth thinking about using the other patch for controller wide formatting of messages? 

If other people feel strongly I'm fine with either.

**Comment by cgwalters** [comment]
OK I pushed the smaller patch for now.

**Comment by cgwalters** [comment]
/test e2e-gcp-op

So let's see if it was just a timeout

**Comment by sinnykumari** [comment]
interesting gcp-op test timeout. Retesting
/retest

**Comment by sinnykumari** [comment]
After looking at MCC log, I would say we should go for the first logging patch added in https://github.com/openshift/machine-config-operator/pull/3135#issuecomment-1128170771 . It provides better readability for drain related log because log starts with nodename.

**Comment by cgwalters** [comment]
OK I replaced with the first patch and also bumped the e2e timeouts to test.
/test e2e-gcp-op

**Comment by sinnykumari** [comment]
For e2e-gcp-op and gcp-op-single-node, drain logs looks fine, they are hitting global timeout. 
gcp-op increased time can be related to change in drain logic. Until we work on splitting these e2e test, I think we can increase these test timeout.

**Review by sinnykumari — APPROVED** [review]
Overall LGTM.

**Comment by yuqi-zhang** [comment]
Hi all, thanks for taking a look at this!

@cgwalters looking at the commits, I can re-format my commit message and lift WIP. Is there anything else I can do?

**Comment by cgwalters** [comment]
> @cgwalters looking at the commits, I can re-format my commit message and lift WIP. Is there anything else I can do?

Sounds good to me!

**Comment by yuqi-zhang** [comment]
Done! The remaining pieces are:

> One option is to make this code dynamically switchable. Jerry and I were chatting about this, I think it's basically we have a global config option (a configmap? annotation on something?; maybe strawman oc create -n openshift-machine-config-operator configmap centralized-drain) that enables this (or disables it) - this reduces the risk of landing. To implement this the operator sync.go code updates the RBAC and also the daemonset to pass e.g. --centralized-drain in between reading manifest and applying.

**Comment by sinnykumari** [comment]
> Done! The remaining pieces are:
> 
> > One option is to make this code dynamically switchable. Jerry and I were chatting about this, I think it's basically we have a global config option (a configmap? annotation on something?; maybe strawman oc create -n openshift-machine-config-operator configmap centralized-drain) that enables this (or disables it) - this reduces the risk of landing. To implement this the operator sync.go code updates the RBAC and also the daemonset to pass e.g. --centralized-drain in between reading manifest and applying.

Are we planning to make this drain behavior configurable as part of the fix?
If I understand correctly, this would require again lot of code change to handle both old and new behavior?

File: pkg/daemon/daemon.go [hunk 10]
```diff
@@ -1721,12 +1658,31 @@ func (dn *Daemon) prepUpdateFromCluster() (*mcfgv1.MachineConfig, *mcfgv1.Machin
...
-	if err := dn.cordonOrUncordonNode(false); err != nil {
-		return err
+	if err := dn.nodeWriter.SetDesiredDrainer(fmt.Sprintf("%s-%s", "uncordon", desiredConfigName)); err != nil {
+		return fmt.Errorf("Could not set drain annotation: %w", err)
+	}
+
+	if err := wait.Poll(10*time.Second, 10*time.Minute, func() (bool, error) {
+		node, getErr := dn.kubeClient.CoreV1().Nodes().Get(context.TODO(), dn.name, metav1.GetOptions{})
+		if getErr != nil {
+			return false, fmt.Errorf("Could not poll node status. Assuming something went wrong")
```
Comment (cgwalters) [inline pkg/daemon/daemon.go:1668]: In general, we really should avoid turning transient network blips into any kind of fatal failure.  OK looking through things I think this will not be actually fatal, we should retry.

But it's not good for us to swallow the error; will do a quick patch.

**Comment by yuqi-zhang** [comment]
> Yeah it would definitely complicate things. I put it out there as a suggestion but...I think I'd vote instead to try to get some decent pre-merge testing (including some manual testing?) here and instead just try landing this.

This PR is definitely a good candidate for payload testing, so let's try that

File: pkg/controller/drain/drain_controller.go (hunk 4 above)
Comment (cgwalters) [reply pkg/controller/drain/drain_controller.go:293]: OK.  Seems fixable by serializing the start time into another annotation or so, but not a blocker.

This all said, I think this "drain killing itself while also draining other nodes" is definitely something to exercise explicitly in manual (and ideally automated tests).

I realized it's simple to do via e.g.:

- Create MC for worker pool
- `oc -n openshift-machine-config-operator delete pod/machine-config-controller-xyz`

(Of course, this is a more synthetic version of the real flow of actually rolling the master pool, but I think should hit the relevant edge cases while being a lot cheaper to test and easier to precisely time the termination, etc.)

Comment (yuqi-zhang) [reply pkg/controller/drain/drain_controller.go:293]: I think another good test to exercise alongside this would be to have an in-progress update on the worker pool, start a second update, delete said pod, and see if the new controller recovers.

And, hmm, maybe recovery from degrade testing, and timeout behaviour, although those are maybe harder to do in a e2e context.

If we can confirm all 3 works, I think we should be good to go in terms of feature parity with what we have (minus some additional slowdowns)

**Comment by sinnykumari** [comment]
> /payload 4.11 ci blocking

This is interesting, didn't know you can run these test from a PR

**Comment by sinnykumari** [comment]
Ran some local test and so far didn't see any unexpected behavior.

**Comment by sinnykumari** [comment]
Let's get this PR merged? Or does this need more testing/review?

**Comment by cheesesashimi** [comment]
Between the payload tests and our passing e2e tests, I think this is ready to merge.

**Comment by yuqi-zhang** [comment]
> I noticed we didn't integrate @cgwalters patches:
https://github.com/cgwalters/machine-config-operator/commit/9aabf12d33e126eb38303d6a49158bfdef130bbf
https://github.com/cgwalters/machine-config-operator/commit/f7dc665c6283bcce2f6a0d7758f5229a44848390
Maybe we should integrate those?

I can add these in

Looks like the only failing test in the payload run was e2e-serial, which failed due to high churn in etcd. Not sure if related or not, but if we get this in early, I think we can get more signal, so a tentative +1 from me as well (after I add the other commits)

**Comment by yuqi-zhang** [comment]
Rebased + added colin's patches

**Comment by sinnykumari** [comment]
Thanks Colin, Jerry and Zack for working on the fix! Team work :)
Let's get this in and have soaking to uncover any unknown issues.
/lgtm
